cli/pack/common.go

```diff
@@ -30,6 +31,40 @@ var (
 	}
 )
 
+func GetTarantoolMinVersion(version string, packType string) (string, error) {
+	var err error
+	var current, border *goVersion.Version
+
+	if border, err = goVersion.NewVersion("2.9.0-alpha"); err != nil {
+		return "", err
+	}
+
+	if current, err = goVersion.NewVersion(version); err != nil {
+		return "", err
+	}
+
+	// Starting from 2.9.0 (since the introduction of new versioning),
+	// the version format has changed.
+	// It needs to be converted to the format in RPM / DEB packages.
+	if current.GreaterThanOrEqual(border) {
+		// We have to convert this format (for example):
+		// 2.10.0-beta1-0-g7da4b1438
+		// To this:
+		// 2.10.0~beta1-1
+		replaced := strings.Replace(version, "-", "~", 1)
+		replaced = strings.SplitN(replaced, "-", 2)[0]
+		return fmt.Sprintf("%s-1", replaced), nil
+	}
+
+	// RPM format
+	if packType == RpmType {
+		return strings.SplitN(version, "-", 2)[0], nil
+	}
+
+	// Deb format
+	return version, nil
+}
```

Comment:
Let me describe the idea behind the function. A developer working on an application on its Linux or Mac OS host. The develover verifies the application on locally installed tarantool version and want to be sure that all required tarantool functionality will be available, when a client installs the application from an RPM or Deb package. We assume that the typical wish is to ensure that tarantool version on the target system is not older than one installed on the developer host. Correct?

The function should return a string (based on `tarantool --version` output), which is suitable to place into the `Requires: tarantool >= {}` the RPM Spec of the application and `Depends: tarantool (>= {})` in the `debian/control` file. Right?

We can construct a table with variants of `tarantool --version` output and RPM/Deb package versions (upstream versions, without distribution's `-1`, `-1.el7` and so on).

| Tarantool version/build | tarantool --version                      | RPM version             | Deb version                 |
| ----------------------- | ---------------------------------------- | ----------------------- | --------------------------- |
| <= 2.8\*                | `X.Y.Z-N-g<commit id>`                   | `X.Y.Z.N`               | `X.Y.Z.N.g<commit id>`      |
| >= 2.10\* (release)     | `X.Y.Z-0-g<commit id>`                   | `X.Y.Z`                 | `X.Y.Z`                     |
| >= 2.10\* (pre-release) | `X.Y.Z-rcN-0-g<commit id>`[^1]           | `X.Y.Z~rcN`[^1]         | `X.Y.Z~rcN`[^1]             |
| >= 2.10\* (nightly)     | `X.Y.Z[-rcN]-N-g<commit id>-dev`[^2][^3] | `X.Y.Z[~rcN].N.dev`[^2] | `X.Y.Z[~rcN].N.dev`[^2][^4] |

[^1]: The same for `alpha` and `beta`.
[^2]: The same for `alpha`, `beta` and the special `entrypoint` mark.
[^3]: It lacks `-dev` suffix until https://github.com/tarantool/tarantool/issues/6183 will be fixed.
[^4]: It also may be meaningful to add `<commit id>` into developer builds of Debian packages. I'm not sure but AFAIR unreleased versions usually have it (unlike releases). The question is to be investigated.

When Tarantool Enterprise is used on a developer host, `cartridge pack` copies tarantool's binary file into the package and don't set the dependency on tarantool. Anyway, if we want to keep `tarantool --version` parsing code universal, it would be good to support parsing of Tarantool Enterprise versions as well. The version is constructed as `<opensource_version>-rM[-macos]` (the `-macos` suffix is optional). The `<opensource_version>` is the `describe --long --abbrev=9 <commit id>` output. `M` is commits count from the sdk repository root to given commit (passed from the sdk repository to the tarantool enterprise build scripts).

<details>
<summary>The similar table for Tarantool Enterprise</summary>

| Tarantool version/build | tarantool --version                 | The same on Mac OS                            |
| ----------------------- | ----------------------------------- | --------------------------------------------- |
| <= 2.8\*                | `X.Y.Z-N-g<commit id>-rM`           | `X.Y.Z-N-g<commit id>-rM-macos`               |
| >= 2.10\* (release)     | `X.Y.Z-0-g<commit id>-rM`           | `X.Y.Z-0-g<commit id>-rM-macos`               |
| >= 2.10\* (pre-release) | `X.Y.Z-rcN-0-g<commit id>-rM`       | `X.Y.Z-rcN-0-g<commit id>-rM-macos`           |
| >= 2.10\* (nightly)     | `X.Y.Z[-rcN]-N-g<commit id>-rM`[^5] | `X.Y.Z[-rcN]-N-g<commit id>-rM-macos`[^5]     |

[^5]: There is no sense in `-dev` for tarantool enterprise, because there is `-rM`.

Notes:

* Tarantool Enterprise is NOT distributed as RPM/Deb package.
* Footnotes from the opensource table are applicable here.
</details>

Details about version relationship operators for RPM and Deb packages (sources: [1][1], [2][2], [3][3]):

1. We can use either upstream version specifier (like `X`, `X.Y`, `X.Y.Z`, `X.Y.Z~beta1`) or full package version (like `X.Y.Z-1`, `X.Y.Z.N-1`, `X.Y.Z-1.el7`).
2. The version is like a border, so we can use, say, `Requires: tarantool >= 1.10.3` despite the fact that tarantool (mostly) has four digit version.

[1]: http://ftp.rpm.org/max-rpm/s1-rpm-depend-manual-dependencies.html
[2]: https://www.debian.org/doc/debian-policy/ch-relationships.html
[3]: https://www.debian.org/doc/debian-policy/ch-controlfields.html#s-f-version

I propose the following rules to form the `>=` dependency:

| Tarantool version/build    | tarantool --version                   | Dependency                       |
| -------------------------- | ------------------------------------- | -------------------------------- |
| <= 2.8\*                   | `X.Y.Z-N-g<commit id>`                | `>= X.Y.Z`                       |
| >= 2.10\* (release)        | `X.Y.Z-0-g<commit id>`                | `>= X.Y.Z`                       |
| >= 2.10\* (pre-release)    | `X.Y.Z-rcN-0-g<commit id>`            | `>= X.Y.Z~rcN`                   |
| >= 2.10\* (nightly, usual) | `X.Y.Z[-rcN]-N-g<commit id>-dev`      | require `--deps` / `--deps-file` |
| >= 2.10\* (nightly, early) | `X.Y.Z-entrypoint-N-g<commit id>-dev` | require `--deps` / `--deps-file` |

The proposal is short: use three upstream version numbers (plus alpha/beta/rc suffix) to point a release. For release build: just copy `X.Y.Z`. For a pre-release: copy `X.Y.Z-rcN`.

For a developer build we can make a guess that everything that is necessary is in the latest (pre-)release (`X.Y.Z-N-g<commit id>-dev` -> `>= X.Y.Z`, `X.Y.Z-rcN-N-g<commit id>-dev` -> `X.Y.Z-rcN`). But I propose another behaviour, because there is a special case.

The corner case is the following: `X.Y.Z-entrypoint-N-g<commit id>-dev` -> `?`. There is no (pre-)release in this feature set, it is unclear whether there are previous bugfix releases. I propose to fail in the case if no tarantool dependency is given thought `--deps` or `--deps-file`. I think we should do that for all developer builds.

NB: Don't forget to re-check that if tarantool dependency is set via `--deps` / `--deps-file`, it will not be overwritten.

I think that it would be better to parse `tarantool --version` into a structure and form the dependency using the structure content. It would be much more easy to spot the idea rather than get it from code based on string split/join/replace operations.

The structure could be the following:

* Major
* Minor
* Patch
* PrereleaseSuffix (optional)
* Revision
* CommitId
* EnterpriseSdkRevision (optional)
* EnterpriseIsOnMacOS (optional)
* IsDeveloperBuild (optional)

So we'll construct `Major.Minor.Patch + (PrereleaseSuffix ? '~' + PrereleaseSuffix : '')`. In fact, we don't even need any custom logic for old / new policy, just process pre-release suffixes correctly.

I'm not insist on 'full parse and re-construction' approach, but currect code looks obscure for me. I'll share several points around current code.

----

Notes on current implementation:

1. It receives the `version` string and it is not clear whether it is full `tarantool --version` part after 'Tarantool' (or 'Tarantool Enterprise') or some part. In fact, it was changed recently.
2. It produces versions with different number of digits for RPM / Deb and I don't understand why.
3. It adds `-1` and it is not clear why.
4. It replaces a first `-` with `~`, but it'll break, say, `X.Y.Z-N-g<commit id>` (when you have no a pre-release suffix).
5. The `2.9.0-alpha` border is unclear for me. Whether it is sorted before or after `2.9.0`? Should not it have tilda instead of dash? Why `2.9` if we started the new policy from `2.10`?

----

Regarding testing.

I would add a unit test with different variants of `tarantool --version` output. See 6b89c1ffebbd39ec5f3d0e5e19a1200c5386239d for examples or, better, construct more real-like examples using the tables above.